More type-safe comparisons in site/templates

frankmayer opened the pull request:
### Summary of Changes
- More type-safe comparisons in site/templates

The changes in this PR should be fairly easy to review. They are only type safe `int` or `bool` comparisons. This PR is done, so the batch PR #12233 can become a little lighter and easier to review. In hope that this will get merged quickly so further work can be done without conflicting with other PRs. ;)

### Testing Instructions

None, should not change behavior


### Documentation Changes Required

None.

frankmayer committed:
More type-safe comparisons in site/templates

bertmert commented:
Could you please explain why we need all these type-safe comparisons?

frankmayer commented:
The quick points:
- Good coding practice
- More strict typing = less error prone code
- Faster processing of such comparisons

There is lots of info on the web on this matter.

mbabker commented:
In general, type safety should be preferred, otherwise you run into a lot of funny PHP quirks.

Essentially, `'0' == 0 == false == null` in PHP with loose typing, same with `'1' == 1 == true`.

Newer PHP versions also start spitting errors when you give non-numeric types (only integer and float are numeric) to be processed for arithmetic operations, older versions were a little less restrictive and would try to do math on a string representation on the value.

andrepereiradasilva commented on templates/beez3/html/com_contact/categories/default_items.php:
```diff
@@ -9,7 +9,7 @@
 defined('_JEXEC') or die;
 $class = ' class="first"';
-if (count($this->items[$this->parent->id]) > 0 && $this->maxLevelcat != 0) :
+if (count($this->items[$this->parent->id]) > 0 && $this->maxLevelcat !== 0) :
```
are we sure maxLevelcat and maxLevel (below) are always an int?

frankmayer replied:
If I am not mistaken, they are both OK. Definitions:
- `maxLevelCat`  here: https://github.com/joomla/joomla-cms/blob/staging/libraries/legacy/view/categories.php#L89
- `maxLevel` here: https://github.com/joomla/joomla-cms/blob/staging/libraries/legacy/view/category.php#L195

frankmayer replied:
But we should probably type cast those...

andrepereiradasilva replied:
the question is, for instance, `$params->get('maxLevelcat', PHP_INT_MAX);` could return a string right?
i would prefer the first alternative too

frankmayer replied:
Yes, unfortunately it could. 
I just checked with the JSON data in the menu table and both are strings!!! Why?

frankmayer replied:
Did a little digging, and it turns out both of them are set in jForm XML files and only contain numeric values (in strings).

As a solution, I propose to type cast when they are fetched from the db: 
```
$maxLevel         = (int) $params->get('maxLevel', -1);
```

andrepereiradasilva replied:
Would that be B/C? Not sure. Here is the B/C politcy
Also, IMHO, to preserve consistency across the core that would be need to be done across all core int fields in all xml forms and all probaly db tables.
So that would probably need to go to 4.0., not sure tough.

frankmayer replied:
I have already done some thinking on that. Will do some prototyping in the next days.

andrepereiradasilva replied:
@frankmayer any news on this?

frankmayer replied:
Hi @andrepereiradasilva, thanks for the code reviews. Hopefully we can get all the remaining improvements merged this time around :smile:
As for your question: I never came around to do some more serious thinking and prototyping. But I agree with you, this might be better suited for 4.0. I will probably reverse this change for now. I think i this should be resolved in a new PR.

frankmayer replied:
@andrepereiradasilva OK, I reversed those strict comparisons for those properties in all related files of this PR. Thanks for pointing that out.

rdeutz committed:
Merge branch 'staging' into more-type-safe-comparisons-in-site-templates

Quy commented on templates/beez3/html/com_contact/category/default.php:
```diff
@@ -35,7 +35,7 @@
 <?php echo $this->loadTemplate('items'); ?>
 
-<?php if (!empty($this->children[$this->category->id])&& $this->maxLevel != 0) : ?>
+<?php if (!empty($this->children[$this->category->id])&& $this->maxLevel !== 0) : ?>
```
Add a space before `&&`.

Quy commented on templates/beez3/html/com_newsfeeds/category/default.php:
```diff
@@ -37,7 +37,7 @@
 <?php echo $this->loadTemplate('items'); ?>
 
-<?php if (!empty($this->children[$this->category->id])&& $this->maxLevel != 0) : ?>
+<?php if (!empty($this->children[$this->category->id])&& $this->maxLevel !== 0) : ?>
```
Add a space before `&&`.

Quy commented on templates/beez3/html/com_weblinks/category/default.php:
```diff
@@ -35,7 +35,7 @@
 <?php endif; ?>
 <?php echo $this->loadTemplate('items'); ?>
-<?php if (!empty($this->children[$this->category->id])&& $this->maxLevel != 0) : ?>
+<?php if (!empty($this->children[$this->category->id])&& $this->maxLevel !== 0) : ?>
```
Add a space before `&&`.

Quy commented on templates/beez3/index.php:
```diff
@@ -19,7 +19,7 @@
 $showleft        = ($this->countModules('position-4') or $this->countModules('position-7') or $this->countModules('position-5'));
 
-if ($showRightColumn == 0 and $showleft == 0)
+if ($showRightColumn === false and $showleft === false)
```
Change `and` to `&&`.

frankmayer committed:
Merge branch 'staging' into more-type-safe-comparisons-in-site-templates

frankmayer committed:
Changes according to reviewer's suggestions

frankmayer committed:
Merge remote-tracking branch 'origin/more-type-safe-comparisons-in-site-templates' into more-type-safe-comparisons-in-site-templates

frankmayer committed:
Merge branch 'staging' into more-type-safe-comparisons-in-site-templates

frankmayer committed:
Merge branch 'staging' into more-type-safe-comparisons-in-site-templates

frankmayer committed:
Additional changes according to reviewer's comments:
Reversed some strict comparisons, as the values compared might not be of same type

frankmayer committed:
Additional changes according to reviewer's comments:
Reversed some more strict comparisons, as the values compared might not be of same type

andrepereiradasilva commented:
I have tested this item :white_check_mark: successfully on 2e806ed12afe0e5932626ab4489aac0dde2da94d<br /><br />On code review<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13310">issues.joomla.org/tracker/joomla-cms/13310</a>.</sub>

Quy commented:
I have tested this item :white_check_mark: successfully on 2e806ed12afe0e5932626ab4489aac0dde2da94d<br /><br />Code review.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13310">issues.joomla.org/tracker/joomla-cms/13310</a>.</sub>

franz-wohlkoenig commented:
RTC after two successful tests.
